process PREPPAINTOR_splitlocus {
    '''
    This process splits a GWAS file into locus files based on lead SNPs and using parameters provided below.

    Inputs
    gwasFile: Path to the GWAS input file
    pvalue_lead: P-value threshold to use for lead variants
    pvalue_nonlead: P-value threshold to use for non-lead variants
    kb: The distance threshold (in kilobases) for lead and non-lead variants
    pvalue_header: Column header for the P-value in the GWAS file
    stderr_header: Column header for the standard error in the GWAS file
    effect_header: Column header for the effect size in the GWAS file
    chromosome_header: Column header for the chromosome in the GWAS file
    effectallele_header: Column header for the effect allele in the GWAS file
    altallele_header: Column header for the alternate allele in the GWAS file
    position_header: Column header for the genomic position in the GWAS file
    rsid_header: Column header for the variant identifier
    zheader_header: Column header for the z-score in the locus files

    Outputs
    Multiple locus-specific files, generated by the main_V2.py script in the output directory 
    specified by params.outputDir_locus.
    
    Script main_V2.py
    The script first creates the output directory if it does not exist. 
    It then runs a Python script main_V2.py with several command line arguments that include 
    the input file, column headers, P-value and other thresholds, and output directory. 
    The main_V2.py script is expected to split the input file into locus-specific files based 
    on the provided parameters and write them to the output directory.
    '''

    publishDir '.', mode: 'copy'

    input:
        path gwasFile
        val pvalue_lead
        val pvalue_nonlead
        val kb
        val pvalue_header
        val stderr_header
        val effect_header
        val chromosome_header
        val effectallele_header
        val altallele_header 
        val position_header
        val rsid_header
        val zheader_header
        

    output:
        path "$params.outputDir_locus/*"

    script:
    """
        mkdir -p ${params.outputDir_locus}
        main_V2.py \\
        --data $gwasFile  \\
        --separator '\t' \\
        --pvalue-header $pvalue_header \\
        --stderr $stderr_header \\
        --effect  $effect_header \\
        --chromosome $chromosome_header \\
        --effect-allele $effectallele_header \\
        --alt-allele  $altallele_header \\
        --position $position_header \\
        --rsid $rsid_header \\
        --Zheader $zheader_header \\
        --kb $kb \\
        --pv-lead $pvalue_lead \\
        --pv-nonlead $pvalue_nonlead \\
        --od ${params.outputDir_locus}
    """
}